plutus-ledger-constraints/src/Ledger/Constraints/ValidityInterval.hs

```diff
@@ -0,0 +1,78 @@
+{-# LANGUAGE DeriveAnyClass     #-}
+{-# LANGUAGE DeriveGeneric      #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts   #-}
+{-# LANGUAGE FlexibleInstances  #-}
+{-# LANGUAGE LambdaCase         #-}
+{-# LANGUAGE MonoLocalBinds     #-}
+{-# LANGUAGE NamedFieldPuns     #-}
+{-# LANGUAGE NoImplicitPrelude  #-}
+{-# LANGUAGE OverloadedStrings  #-}
+
+module Ledger.Constraints.ValidityInterval where
+
+import Data.Aeson (FromJSON, ToJSON)
+import GHC.Generics (Generic)
+import Plutus.V1.Ledger.Interval (Extended (Finite, NegInf, PosInf), Interval (Interval), LowerBound (LowerBound),
+                                  UpperBound (UpperBound))
+import PlutusTx.Prelude (Bool (False, True), Enum (succ), Functor (fmap), Maybe (Just, Nothing))
+import Prelude qualified as Haskell
+
+-- | 'ValidityInterval' is a half open interval. Closed (inclusive) on the bottom, open
+-- (exclusive) on the top. A 'Nothing' on the bottom is negative infinity, and a 'Nothing'
+-- on the top is positive infinity.
+data ValidityInterval a = ValidityInterval
+  { invalidBefore    :: !(Maybe a) -- ^ Inclusive lower bound or negative infinity
+  , invalidHereafter :: !(Maybe a) -- ^ Exclusive upper bound or positive infinity
+  }
+  deriving stock (Haskell.Show, Generic, Haskell.Eq)
+  deriving anyclass (ToJSON, FromJSON)
+
+instance Functor ValidityInterval where
+  fmap f (ValidityInterval from' to') = ValidityInterval (fmap f from') (fmap f to')
+
+{-# INLINABLE interval #-}
+-- | @interval a b@ includes all values that are greater than or equal to @a@
+-- and smaller than @b@. In math. notation: [a,b)
+interval :: a -> a -> ValidityInterval a
+interval s s' = ValidityInterval (Just s) (Just s')
+
+{-# INLINABLE from #-}
+-- | @from a@ is an 'ValidityInterval' that includes all values that are
+--  greater than or equal to @a@. In math. notation: [a,+∞]
+from :: a -> ValidityInterval a
+from s = ValidityInterval (Just s) Nothing
+
+{-# INLINABLE to #-}
+-- | @to a@ is an 'ValidityInterval' that includes all values that are
+--  smaller than to @a@. In math. notation: [-∞,a)
+to :: a -> ValidityInterval a
+to s = ValidityInterval Nothing (Just s)
```

Comment:
The behaviour is aligned with the documentation but I wonder whether the given upper bound should be included in the interval or not: when we count _to_ a given value, we usually include this value.